pkg/endpoint/events.go

```diff
@@ -30,7 +30,23 @@ func (ev *EndpointRegenerationEvent) Handle(res chan interface{}) {
 	e := ev.ep
 	regenContext := ev.regenContext
 
-	err := e.rlockAlive()
+	// Compute policy before acquiring the build permit in
+	// QueueEndpointBuild below
+	err, release := e.ComputeInitialPolicy(regenContext)
```

Comment:
I think it would be helpful to state here that in the happy path, after returning from `ComputeInitialPolicy()`, the `buildMutex` is still held, so that the reader knows immediately, without having to inspect the calling code.

Reply:
it is not held:
```
func (e *Endpoint) ComputeInitialPolicy(regenContext *regenerationContext) (error, func()) {
	stats := &regenContext.Stats
	datapathRegenCtxt := regenContext.datapathRegenerationContext

	// buildMutex needed for all policy computation functions below.
	e.buildMutex.Lock()
	defer e.buildMutex.Unlock()
...
```

Reply:
But the handler serializes build events!

Reply:
And the policy is held in the `regenContext` created for the event.

So only one build event can be in progress for each endpoint, so there is no possibility for a "policy race" even if there would be 2 builds requested. The 2nd one will stay in the queue. Note that `QueueEndpointBuild` does not actually queue, it waits for a the `daemon.buildEndpointSem`.